### PR TITLE
RobotFactoryPosition: Infrastructure and Restarts

### PR DESCRIPTION
**Pull request by sid-parikh**
## Description
Updates positions based on game state (e.g. Penalty, Kickoff, Free Kick)

## Associated / Resolved Issue
Resolves # or [ClickUp card](https://app.clickup.com/t/86azbvek6)
Resolves # or [ClickUp card](https://app.clickup.com/t/86aznadc5)

## Steps to Test
Use external ref to call for kickoffs, penalties, and free kicks for both teams and verify rule-compliance.

## Key Files to Review
 * robot_factory_position.cpp
 * robot_factory_position.hpp
 * offense.cpp
 * offense.hpp
 * penalty_non_kicker.cpp
 * penalty_non_kicker.hpp
 * penalty_player.cpp
 * penalty_player.hpp
 * smartidling.cpp
 * smartidling.hpp
 * free_kicker.cpp
 * free_kicker.hpp

## Review Checklist

- [ ] **Docstrings**: All methods and classes should have the file appropriate docstrings which follow the guidelines in the ["Contributing" page](https://rj-rc-software.readthedocs.io/en/latest/contributing.html) of our docs.
- [ ] **Remove extra print statements**: Any print statements used for debugging should be removed
- [ ] **Tag reviewers**: Tag some people for review and ping them on Slack

## (Optional) Sub-issues (for drafts)
_Note: if you find yourself breaking this PR into many smaller features, it may make sense to break up the PR into logical units based on these features._
- [ ] Step 1
- [ ] Step 2

### CODE REVIEW

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp [hunk 1]
```diff
@@ -1,37 +1,233 @@
...
+        set_current_position<Goalie>();
+        return current_position_->get_task(*last_world_state_, field_dimensions_,
+                                           current_play_state_);
     }
 
-    // TODO (Rishi and Jack): Make this synchronized across all robots to avoid race conditions
+    // Update our state
+    process_play_state();
+
+    update_position();
```
Comment (Squid5678) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp:32]: Comment here? Might be useful.

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp [hunk 2]
```diff
@@ -1,37 +1,233 @@
...
+                             }
+                             return a.second < b.second;
+                         });
 
+    // Closest is an iterator to the pair (robot_id, distance)
+    return closest->first == robot_id_;
+}
+
+void RobotFactoryPosition::set_default_position() {
+    // TODO (Rishi and Jack): Make this synchronized across all robots to avoid race conditions
```
Comment (Squid5678) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp:219]: Is this still in progress?

File: soccer/src/soccer/planning/planner/collect_path_planner.cpp [hunk 3]
```diff
@@ -110,6 +110,14 @@ Trajectory CollectPathPlanner::plan(const PlanRequest& plan_request) {
     std::vector<DynamicObstacle> dynamic_obstacles;
     fill_obstacles(plan_request, &static_obstacles, &dynamic_obstacles, false);
 
+    // Return an empty trajectory if the ball is hitting static obstacles
```
Comment (Squid5678) [inline soccer/src/soccer/planning/planner/collect_path_planner.cpp:113]: Why is this still here?

File: soccer/src/soccer/strategy/agent/communication/seeker_request.hpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    static strategy::communication::SeekerRequest from_ros(
+        const rj_msgs::msg::SeekerRequest& from) {
+        return strategy::communication::SeekerRequest{
+            from.request_uid,     from.robot_id, from.seeking_point_x,
+            from.seeking_point_y, from.adding,
+        };
+    }
+};
+
+ASSOCIATE_CPP_ROS(strategy::communication::SeekerRequest, rj_msgs::msg::SeekerRequest);
```
Comment (Squid5678) [inline soccer/src/soccer/strategy/agent/communication/seeker_request.hpp:50]: Curious on what this line does in particular

File: soccer/src/soccer/strategy/agent/communication/communication.hpp
Comment (jvogt23) [inline soccer/src/soccer/strategy/agent/communication/communication.hpp:1]: Is there another way to do the "if else" blocks in here, specifically on the RosConverter struct? I'm mostly just curious.

File: soccer/src/soccer/strategy/agent/position/free_kicker.cpp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        this->field_dimensions_.their_goal_loc().x() + 0.5 * this->field_dimensions_.goal_width(),
+        this->field_dimensions_.their_goal_loc().y()};
+
+    planning::LinearMotionInstant target{goal_corner};
+    auto line_kick_cmd = planning::MotionCommand{"line_kick", target};
+    intent.motion_command = line_kick_cmd;
+
+    // note: the way this is set up makes it impossible to
+    // shoot on time without breakbeam
+    // TODO(Kevin): make intent hold a manip msg instead? to be cleaner?
```
Comment (Squid5678) [inline soccer/src/soccer/strategy/agent/position/free_kicker.cpp]: Still in progress?

File: soccer/src/soccer/strategy/agent/communication/kicker_request.hpp
Comment (petergarud) [inline soccer/src/soccer/strategy/agent/communication/kicker_request.hpp:1]: Makes sense to create this class for communication. Good job!

**Review by Squid5678 — APPROVED** [review]
Good job

File: soccer/src/soccer/strategy/agent/communication/seeker_request.hpp
Comment (petergarud) [inline soccer/src/soccer/strategy/agent/communication/seeker_request.hpp:1]: Good job at modeling how other request was written to maintain code style and readibility.

File: soccer/src/soccer/strategy/agent/position/penalty_non_kicker.cpp
Comment (petergarud) [inline soccer/src/soccer/strategy/agent/position/penalty_non_kicker.cpp:1]: I'm assuming this is smart idle behavior

File: soccer/src/soccer/strategy/agent/position/defense.cpp [hunk 6]
```diff
@@ -4,8 +4,16 @@ namespace strategy {
 
 Defense::Defense(int r_id) : Position(r_id, "Defense"), marker_{field_dimensions_} {}
 
+Defense::Defense(const Position& other) : Position{other}, marker_{field_dimensions_} {
+    SPDLOG_INFO("HELLO {}", robot_id_);
```
Comment (jvogt23) [inline soccer/src/soccer/strategy/agent/position/defense.cpp]: Does the SPDLOG command need to be here? If so, possibly make it more descriptive.

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp [hunk 7]
```diff
@@ -99,6 +290,17 @@ void RobotFactoryPosition::receive_communication_response(
 
 communication::PosAgentResponseWrapper RobotFactoryPosition::receive_communication_request(
     communication::AgentPosRequestWrapper request) {
+    if (const communication::KickerRequest* kicker_request =
+            std::get_if<communication::KickerRequest>(&request.request)) {
+        if (kicker_distances_.size() >= 1 && !have_all_kicker_responses() &&
+            current_position_->get_name() != "GoalKicker") {
+            bool prev = kicker_distances_.count(kicker_request->robot_id) >= 1;
+            if (!prev) {
```
Comment (petergarud) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp:299]: No more infinite request! Great!

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp
Comment (petergarud) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp:1]: Looks much more organized and understandable. Good job!

**Review by petergarud — COMMENTED** [review]
Great job overall! I also ran it and it seemed to be working especially the seeking. There were a few cases where the robots started to run into each other though

File: soccer/src/soccer/strategy/agent/position/free_kicker.cpp [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+namespace strategy {
+
+FreeKicker::FreeKicker(int r_id) : Position(r_id, "FreeKicker") {}
+
+FreeKicker::FreeKicker(const Position& other) : Position{other} { position_name_ = "FreeKicker"; }
+
+std::optional<RobotIntent> FreeKicker::derived_get_task(RobotIntent intent) {
+    // Penalty Kicker kicks the ball into the goal
+
+    SPDLOG_INFO("Free Kicker {} is running", this->robot_id_);
```
Comment (jvogt23) [inline soccer/src/soccer/strategy/agent/position/free_kicker.cpp]: I'm not sure if I am a good judge of whether or not a log message is needed, but I'm wondering if there may be enough log messages throughout the code to clutter the logs a bit. Could some of these be commented out?

**Review by jvogt23 — CHANGES_REQUESTED** [review]

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp (hunk 2 above)
Comment (rishiso) [reply soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp:219]: Yes this is a problem for agent 7

File: soccer/src/soccer/planning/planner/collect_path_planner.cpp (hunk 3 above)
Comment (rishiso) [reply soccer/src/soccer/planning/planner/collect_path_planner.cpp:113]: Can you clarify? It makes sense to do this if we happen to hit a static obstacle by chance. Right?

File: soccer/src/soccer/strategy/agent/communication/seeker_request.hpp (hunk 4 above)
Comment (rishiso) [reply soccer/src/soccer/strategy/agent/communication/seeker_request.hpp:50]: It associates the struct with the ROS message

File: soccer/src/soccer/strategy/agent/communication/communication.hpp
Comment (rishiso) [reply soccer/src/soccer/strategy/agent/communication/communication.hpp:1]: No, it uses an assignment inside the condition which can't be done with a switch statement

**Review by petergarud — APPROVED** [review]

**Review by jvogt23 — APPROVED** [review]
idek what i requested changes on

**Pull request merged**